Sources/StreamChat/Utils/LazyCachedMapCollection.swift

```diff
@@ -0,0 +1,80 @@
+//
+// Copyright © 2021 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+/// Read-only collection that applies transformation to element on first access.
+///
+/// Compared to `LazyMapCollection` do not perform transformation on `count` call.
+public struct LazyCachedMapCollection<Element>: RandomAccessCollection {
+    public typealias Index = Int
+
+    public func index(before i: Index) -> Index {
+        cache.values.index(before: i)
+    }
+
+    public func index(after i: Index) -> Index {
+        cache.values.index(after: i)
+    }
+
+    public var startIndex: Index { cache.values.startIndex }
+
+    public var endIndex: Index { cache.values.endIndex }
+
+    public var count: Index { cache.values.count }
+
+    public init<Collection: RandomAccessCollection, SourceElement>(
+        source: Collection,
+        map: @escaping (SourceElement) -> Element
+    ) where Collection.Element == SourceElement {
+        let erased = Array(source)
```

Comment:
If we use `AnyRandomAccessCollection` for type erasure, we need to find way how to map `Int` index to `Index` type of source `Collection`. Otherwise usage of `AnyIndex` leads to runtime error.
I let stdlib to solve it.

Reply:
Is this safe to do here? For example, if the source is some kind of other lazy collection, this step means it will be finalized.